### PR TITLE
Add file extension to untitled files' filenames using VSCode autodetection

### PR DESCRIPTION
**Pull request by LeshemNoa**
Piggybacking VSCode's ML based language autodetection on untitled files, we can provide a better experience on such files.

### CODE REVIEW

File: src/runCompletion.ts [hunk 1]
```diff
@@ -52,3 +53,22 @@ function getMaxResults(): number {
...
+
+export function getFileNameWithExtension(
+  document: TextDocument
+): string {
+  const { languageId } = document;
+  let filenameWithExtension = document.fileName;
+  if (document.isUntitled) {
+    if (isKnownLanguage(languageId)) {
+      filenameWithExtension = document.fileName.concat(languages[languageId]);
+    }
```
Comment (yonip23) [inline src/runCompletion.ts:71]: i'd create a function `getFileExtension(languageId: string): string | undefined` instead of `isKnownLanguage`, so that this code can look like so:

```typescript
const fileExtension = getFileExtension(languageId);
if (fileExtension) {
    filenameWithExtension = document.fileName.concat(fileExtension);
}
```

**Comment by LeshemNoa** [comment]
@dimacodota the file was constructed manually, combining data from
- [the VSCode language identifier list](https://code.visualstudio.com/docs/languages/identifiers) 
- The output returned from `languages.getLanguages`, which surprisingly is not 1-1 with the ID list they documented themselves. Documented [here](https://code.visualstudio.com/api/references/vscode-api#languages)
- Our own `languages.yaml` from acpl.

**Comment by LeshemNoa** [comment]
@yonip23 I don't understand what I did to cause this test case to fail, it seems totally unrelated! 🤡 ʜᴇʟᴘ

File: src/test/suite/untitledFile.test.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    const res = await activate();
+    editor = res?.editor;
+    doc = res?.doc;
+  });
+
+  afterEach(async () => {
+    await vscode.commands.executeCommand("workbench.action.closeActiveEditor");
+  });
+
+  const tableIt = async (input: string, fileExtension: string) => {
```
Comment (yonip23) [inline src/test/suite/untitledFile.test.ts:22]: change to a meaningful name, something like `testUndefinedFileWith`

File: src/test/suite/utils/helper.ts [hunk 3]
```diff
@@ -14,10 +14,12 @@ export type BinaryGenericRequest<T> = {
...
  */
 export async function activate(
-  docUri: vscode.Uri
+  docUri?: vscode.Uri
 ): Promise<{ editor: TextEditor; doc: TextDocument } | null> {
   try {
-    const doc = await vscode.workspace.openTextDocument(docUri);
+    const doc = docUri
+      ? await vscode.workspace.openTextDocument(docUri)
+      : await vscode.workspace.openTextDocument(); // opens an untitled document
```
Comment (yonip23) [inline src/test/suite/utils/helper.ts:22]: I don't think you need to do that, if `docUri` is `undefined`, calling `vscode.workspace.openTextDocument(docUri)` is equivalent to calling it without arguments.

File: src/test/suite/untitledFile.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    editor = editor as vscode.TextEditor;
+    doc = doc as vscode.TextDocument;
+    await editor.edit((edit) => {
+      edit.insert(new vscode.Position(0, 0), input);
+    });
+    await sleep(1000); // wait for editor to detect language
```
Comment (yonip23) [inline src/test/suite/untitledFile.test.ts:28]: isn't there any better way to confirm that the editor has detected the language? tests with `sleep`s are prone to being flaky.

File: src/test/suite/utils/helper.ts (hunk 3 above)
Comment (yonip23) [reply src/test/suite/utils/helper.ts:22]: if `docUri` is undefined, it'd be considered the same as calling the function without any arguments --> the overload will resolve to the first one.

give it a try, i'm pretty sure it'll work.

File: src/test/suite/untitledFile.test.ts (hunk 4 above)
Comment (yonip23) [reply src/test/suite/untitledFile.test.ts:31]: mmm ok then, lets hope it'll be stable enough 🙂

File: src/test/suite/utils/helper.ts (hunk 3 above)
Comment (yonip23) [reply src/test/suite/utils/helper.ts:22]: oh well, then you can resolve this comment.
sorry for bothering you 🙈

**Review by yonip23 — APPROVED** [review]

File: src/test/suite/utils/helper.ts (hunk 3 above)
Comment (assapir) [reply src/test/suite/utils/helper.ts:22]: this is why I hate TS :(

File: src/runCompletion.ts [hunk 5]
```diff
@@ -48,3 +49,23 @@ function getMaxResults(): number {
...
+export function getLanguageFileExtension(
+  languageId: string
+): string | undefined {
+  return languages[languageId as KnownLanguageType];
+}
+
+export function getFileNameWithExtension(document: TextDocument): string {
+  const { languageId, fileName } = document;
+  let filenameWithExtension = fileName;
+  if (document.isUntitled) {
```
Comment (assapir) [inline src/runCompletion.ts:64]: I always prefer early returns, so:
```suggestion
  if (!document.isUntitled) {
       return filenameWithExtension;
   }
```

**Review by dimacodota — APPROVED** [review]
lgtm

File: src/runCompletion.ts [hunk 6]
```diff
@@ -48,3 +49,23 @@ function getMaxResults(): number {
...
+
+export function getLanguageFileExtension(
+  languageId: string
+): string | undefined {
+  return languages[languageId as KnownLanguageType];
+}
+
+export function getFileNameWithExtension(document: TextDocument): string {
+  const { languageId, fileName } = document;
+  let filenameWithExtension = fileName;
```
Comment (assapir) [inline src/runCompletion.ts:63]: why is that needed?

File: src/runCompletion.ts [hunk 7]
```diff
@@ -48,3 +49,23 @@ function getMaxResults(): number {
...
+  if (document.isUntitled) {
+    const extension = getLanguageFileExtension(languageId);
+    if (extension) {
+      filenameWithExtension = fileName.concat(extension);
```
Comment (assapir) [inline src/runCompletion.ts:67]:
```suggestion
      return  fileName.concat(extension);
```

**Review by assapir — APPROVED** [review]

File: src/runCompletion.ts (hunk 6 above)
Comment (LeshemNoa) [reply src/runCompletion.ts:63]: Current binary API infers the file's language using its full name including the file extension. Untitled files have no file extension to their name. That way we can add the file extension manually to the untitled file's name.

**Review by assapir — APPROVED** [review]
Ahla
